File: event_engine.py
"""YAML 기반 이벤트 엔진

YAML 파일에 정의된 스토리 이벤트를 해석·실행한다.
data/areas/*.yaml 파일을 읽어 지역 이벤트를 자동 등록한다.
"""

import os
import random
import yaml

from story_utils import _set_location, _input_choice, _divider, _pause, game_over
from combat import battle, spawn_enemy
from inventory import open_shop, open_merc_shop, show_inventory, equip_item


DATA_DIR = os.path.join(os.path.dirname(__file__), "data", "areas")


# ═══════════════════════════════════════════════════════
#  텍스트 포매팅
# ═══════════════════════════════════════════════════════

def _format_text(text: str, player) -> str:
    """텍스트 안의 {player.xxx} 변수를 치환한다."""
    return text.format(player=player)


# ═══════════════════════════════════════════════════════
#  조건 판정
# ═══════════════════════════════════════════════════════

def _check_condition(cond, player) -> bool:
    """단일 조건 또는 조건 리스트를 평가한다.

    지원 조건:
      flag: "key"                 → story_flags[key] == True
      not_flag: "key"             → story_flags[key] != True
      any_flag: ["a", "b"]        → 하나라도 True
      all_flags: ["a", "b"]       → 전부 True
      has_item: "아이템명"         → 인벤토리에 보유
      not_has_item: "아이템명"     → 인벤토리에 미보유
      equipped_weapon: "무기명"    → 장착 무기 확인
      gold_gte: N                 → gold >= N
      gold_lt: N                  → gold < N
      dark_gte: N                 → dark_points >= N
      dark_lt: N                  → dark_points < N
      dark_lte: N                 → dark_points <= N
      hp_lte: N                   → hp <= N
      level_gte: N                → level >= N
      random_lt: 0.3              → random() < 0.3  (30% 확률)
    """
    if cond is None:
        return True
    if isinstance(cond, list):
        return all(_check_condition(c, player) for c in cond)

    flags = getattr(player, "story_flags", {})

    if "flag" in cond:
        return bool(flags.get(cond["flag"], False))
    if "not_flag" in cond:
        return not flags.get(cond["not_flag"], False)
    if "any_flag" in cond:
        return any(flags.get(f, False) for f in cond["any_flag"])
    if "all_flags" in cond:
        return all(flags.get(f, False) for f in cond["all_flags"])
    if "has_item" in cond:
        return cond["has_item"] in player.inventory
    if "not_has_item" in cond:
        return cond["not_has_item"] not in player.inventory
    if "equipped_weapon" in cond:
        w = player.equipped_weapon
        return w and w.get("name") == cond["equipped_weapon"]
    if "gold_gte" in cond:
        return player.gold >= cond["gold_gte"]
    if "gold_lt" in cond:
        return player.gold < cond["gold_lt"]
    if "dark_gte" in cond:
        return player.dark_points >= cond["dark_gte"]
    if "dark_lt" in cond:
        return player.dark_points < cond["dark_lt"]
    if "dark_lte" in cond:
        return player.dark_points <= cond["dark_lte"]
    if "hp_lte" in cond:
        return player.hp <= cond["hp_lte"]
    if "level_gte" in cond:
        return player.level >= cond["level_gte"]
    if "random_lt" in cond:
        return random.random() < cond["random_lt"]

    return True


# ═══════════════════════════════════════════════════════
#  액션 실행기
# ═══════════════════════════════════════════════════════

def _run_actions(actions: list, player, scenes: dict) -> str | None:
    """액션 리스트를 순차 실행. 엔딩 문자열을 반환하면 즉시 전파."""
    if not actions:
        return None
    for action in actions:
        result = _run_action(action, player, scenes)
        if result is not None:
            return result
    return None


def _run_action(action: dict, player, scenes: dict) -> str | None:
    """단일 액션을 실행한다."""
    act = action.get("action", action.get("type", ""))

    # ─── 텍스트 ───
    if act == "print":
        text = _format_text(action["text"], player)
        print(text)
        return None

    if act == "divider":
        _divider(action.get("title", ""))
        return None

    if act == "pause":
        _pause()
        return None

    # ─── 플래그 ───
    if act == "set_flag":
        val = action.get("value", True)
        player.story_flags[action["key"]] = val
        return None

    # ─── 아이템 ───
    if act == "add_item":
        item = action["item"]
        count = action.get("count", 1)
        for _ in range(count):
            player.inventory.append(item)
        if action.get("message", True):
            print(f"  ▶ {item} x{count}을(를) 획득했습니다!")
        return None

    if act == "remove_item":
        item = action["item"]
        if item in player.inventory:
            player.inventory.remove(item)
        return None

    # ─── 골드 ───
    if act == "add_gold":
        amt = action["amount"]
        player.gold += amt
        if action.get("message", True):
            print(f"  ▶ {amt}G 획득!")
        return None

    if act == "sub_gold":
        amt = action["amount"]
        player.gold = max(0, player.gold - amt)
        if action.get("message", True):
            print(f"  ▶ {amt}G 소비!")
        return None

    # ─── 스탯 ───
    if act == "heal":
        amt = action.get("amount")
        if amt == "full" or amt is None:
            player.hp = player.max_hp
        else:
            player.heal(amt)
        if action.get("message", True):
            print(f"  ▶ HP가 회복되었습니다! ({player.hp}/{player.max_hp})")
        return None

    if act == "damage":
        amt = action["amount"]
        player.hp = max(1, player.hp - amt)
        if action.get("message", True):
            print(f"  ▶ {amt}의 피해를 입었습니다! (HP: {player.hp})")
        return None

    if act == "add_stat":
        for stat_name in ("attack", "defense", "max_hp"):
            if stat_name in action:
                cur = getattr(player, stat_name)
                setattr(player, stat_name, cur + action[stat_name])
                if action.get("message", True):
                    print(f"  ▶ {stat_name} +{action[stat_name]}!")
        return None

    if act == "set_stat":
        for stat_name in ("attack", "defense", "max_hp", "hp"):
            if stat_name in action:
                setattr(player, stat_name, action[stat_name])
        return None

    # ─── 어둠 점수 ───
    if act == "add_dark":
        amt = action.get("amount", 1)
        player.dark_points += amt
        if action.get("message", True):
            print(f"  ▶ 어둠 점수 +{amt}")
        return None

    if act == "sub_dark":
        amt = action.get("amount", 1)
        player.dark_points = max(0, player.dark_points - amt)
        if action.get("message", True):
            print(f"  ▶ 어둠 점수 -{amt}")
        return None

    if act == "set_dark":
        player.dark_points = action["amount"]
        return None

    # ─── 전투 ───
    if act == "battle":
        enemy_key = action["enemy"]
        enemy, key = spawn_enemy(enemy_key)
        if "name" in action:
            enemy.name = action["name"]
        if "hp_multiply" in action:
            enemy.hp = int(enemy.hp * action["hp_multiply"])
            enemy.max_hp = enemy.hp
        if "hp_add" in action:
            enemy.hp += action["hp_add"]
            enemy.max_hp = enemy.hp
        result = battle(player, enemy, key)
        if result == "lose":
            on_lose = action.get("on_lose", [{"action": "game_over"}])
            return _run_actions(on_lose, player, scenes)
        on_win = action.get("on_win", [])
        return _run_actions(on_win, player, scenes)

    # ─── 상점 ───
    if act == "open_shop":
        open_shop(player)
        return None

    if act == "open_merc_shop":
        open_merc_shop(player)
        return None

    # ─── 인벤토리/상태 ───
    if act == "show_inventory":
        show_inventory(player)
        return None

    if act == "equip_item":
        equip_item(player)
        return None

    if act == "show_status":
        player.full_status()
        return None

    # ─── 위치 ───
    if act == "set_location":
        _set_location(player, action["zone"])
        return None

    # ─── 지도 ───
    if act == "show_map":
        from map import show_world_map, show_castle_map
        kind = action.get("kind", "world")
        if kind == "castle":
            explored = action.get("explored")
            show_castle_map(player, explored)
        else:
            show_world_map(player)
        return None

    # ─── 저장 ───
    if act == "save":
        from save_manager import prompt_save
        location = action.get("location", "아르카디아 성당")
        prompt_save(player, location)
        return None

    # ─── 게임 오버 ───
    if act == "game_over":
        return game_over(player)

    # ─── 반환 (엔딩 등) ───
    if act == "return":
        return action["value"]

    # ─── 선택지 메뉴 ───
    if act == "menu":
        return _run_menu(action, player, scenes)

    # ─── 조건 분기 ───
    if act == "if":
        if _check_condition(action["cond"], player):
            return _run_actions(action.get("then", []), player, scenes)
        else:
            return _run_actions(action.get("else", []), player, scenes)

    # ─── 랜덤 분기 ───
    if act == "random":
        return _run_random(action, player, scenes)

    # ─── 씬 이동 ───
    if act == "goto":
        scene_name = action["scene"]
        scene = scenes.get(scene_name)
        if scene:
            return _run_scene(scene, player, scenes)
        return None

    # ─── 반복 ───
    if act == "loop":
        while True:
            result = _run_actions(action.get("body", []), player, scenes)
            if result is not None:
                if result == "__break__":
                    return None
                return result

    if act == "break":
        return "__break__"

    return None


# ═══════════════════════════════════════════════════════
#  메뉴 (선택지)
# ═══════════════════════════════════════════════════════

def _run_menu(action: dict, player, scenes: dict) -> str | None:
    """조건부 동적 메뉴를 실행한다."""
    options_def = action.get("options", [])

    visible = []
    for opt in options_def:
        cond = opt.get("when")
        if cond is not None and not _check_condition(cond, player):
            continue
        visible.append(opt)

    if not visible:
        return None

    for i, opt in enumerate(visible, 1):
        print(f"  {i}. {_format_text(opt['label'], player)}")

    choice = _input_choice(set(range(1, len(visible) + 1)))
    chosen = visible[choice - 1]

    if "goto" in chosen:
        scene = scenes.get(chosen["goto"])
        if scene:
            return _run_scene(scene, player, scenes)
        return None

    return _run_actions(chosen.get("actions", []), player, scenes)


# ═══════════════════════════════════════════════════════
#  랜덤 분기
# ═══════════════════════════════════════════════════════

def _run_random(action: dict, player, scenes: dict) -> str | None:
    branches = action.get("branches", [])
    if not branches:
        return None

    weights = [b.get("weight", 1.0) for b in branches]
    total = sum(weights)
    r = random.random() * total
    cumulative = 0.0
    for branch in branches:
        cumulative += branch.get("weight", 1.0)
        if r < cumulative:
            return _run_actions(branch.get("actions", []), player, scenes)
    return _run_actions(branches[-1].get("actions", []), player, scenes)


# ═══════════════════════════════════════════════════════
#  씬 실행
# ═══════════════════════════════════════════════════════

def _run_scene(scene: dict, player, scenes: dict) -> str | None:
    """씬(scene) 하나를 실행한다. 씬 = 액션 리스트 또는 딕셔너리."""
    if isinstance(scene, list):
        return _run_actions(scene, player, scenes)
    actions = scene.get("actions", [])
    return _run_actions(actions, player, scenes)


# ═══════════════════════════════════════════════════════
#  YAML 파일 로드 + 실행
# ═══════════════════════════════════════════════════════

def load_yaml_event(filepath: str) -> dict:
    """YAML 이벤트 파일을 로드한다."""
    with open(filepath, "r", encoding="utf-8") as f:
        return yaml.safe_load(f)


def run_yaml_event(player, zone: str) -> str | None:
    """data/areas/{zone}.yaml를 로드하여 실행한다."""
    filepath = os.path.join(DATA_DIR, f"{zone}.yaml")
    if not os.path.exists(filepath):
        return None

    data = load_yaml_event(filepath)
    event = data.get(zone, data)

    scenes = event.get("scenes", {})
    main_actions = event.get("actions", [])

    return _run_actions(main_actions, player, scenes)


# ═══════════════════════════════════════════════════════
#  자동 등록 (area_registry 연동)
# ═══════════════════════════════════════════════════════

def register_yaml_events():
    """data/areas/ 내 모든 YAML 파일을 스캔하여 이벤트를 자동 등록한다."""
    from area_registry import register_event

    if not os.path.isdir(DATA_DIR):
        return

    for filename in os.listdir(DATA_DIR):
        if not filename.endswith((".yaml", ".yml")):
            continue
        zone = filename.rsplit(".", 1)[0]

        def make_handler(z):
            def handler(player):
                return run_yaml_event(player, z)
            return handler

        register_event(zone, make_handler(zone))

File: web/js/event_engine.js
/* event_engine.js - YAML 이벤트 엔진 (Python event_engine.py의 JS 포팅) */

const EventEngine = {

  /* ───── 텍스트 포매팅 ───── */

  formatText(text, player) {
    if (!text || typeof text !== 'string') return text || '';
    return text.replace(/\{player\.(\w+)\}/g, (_, prop) => {
      if (prop === 'hp') return player.hp;
      if (prop === 'max_hp' || prop === 'maxHp') return player.maxHp;
      if (prop === 'name') return player.name;
      if (prop === 'job') return player.job;
      if (prop === 'level') return player.level;
      if (prop === 'gold') return player.gold;
      if (prop === 'attack') return player.attack;
      if (prop === 'defense') return player.defense;
      if (prop === 'exp') return player.exp;
      if (prop === 'dark_points' || prop === 'darkPoints') return player.darkPoints;
      return `{player.${prop}}`;
    });
  },

  /* ───── 조건 평가 ───── */

  checkCondition(cond, player) {
    if (cond === null || cond === undefined) return true;

    if (Array.isArray(cond)) {
      return cond.every(c => this.checkCondition(c, player));
    }

    const flags = player.storyFlags;

    if ('flag' in cond)
      return !!flags[cond.flag];
    if ('not_flag' in cond)
      return !flags[cond.not_flag];
    if ('any_flag' in cond)
      return cond.any_flag.some(f => !!flags[f]);
    if ('all_flags' in cond)
      return cond.all_flags.every(f => !!flags[f]);
    if ('has_item' in cond)
      return player.inventory.includes(cond.has_item);
    if ('not_has_item' in cond)
      return !player.inventory.includes(cond.not_has_item);
    if ('equipped_weapon' in cond) {
      const w = player.equippedWeapon;
      return w && w.name === cond.equipped_weapon;
    }
    if ('gold_gte' in cond)
      return player.gold >= cond.gold_gte;
    if ('gold_lt' in cond)
      return player.gold < cond.gold_lt;
    if ('dark_gte' in cond)
      return player.darkPoints >= cond.dark_gte;
    if ('dark_lt' in cond)
      return player.darkPoints < cond.dark_lt;
    if ('dark_lte' in cond)
      return player.darkPoints <= cond.dark_lte;
    if ('hp_lte' in cond)
      return player.hp <= cond.hp_lte;
    if ('level_gte' in cond)
      return player.level >= cond.level_gte;
    if ('random_lt' in cond)
      return Math.random() < cond.random_lt;

    return true;
  },

  /* ───── 액션 리스트 실행 ───── */

  async runActions(actions, player, scenes) {
    if (!actions || !Array.isArray(actions)) return null;
    for (const action of actions) {
      const result = await this.runAction(action, player, scenes);
      if (result !== null && result !== undefined) return result;
    }
    return null;
  },

  /* ───── 단일 액션 실행 ───── */

  async runAction(action, player, scenes) {
    const act = action.action || action.type || '';

    /* ── 텍스트 ── */
    if (act === 'print') {
      const text = this.formatText(action.text, player);
      text.split('\n').forEach(line => UI.addLog(line));
      return null;
    }

    if (act === 'divider') {
      UI.addDivider(action.title || '');
      return null;
    }

    if (act === 'pause') {
      await UI.waitForTap();
      return null;
    }

    /* ── 플래그 ── */
    if (act === 'set_flag') {
      const val = action.value !== undefined ? action.value : true;
      player.storyFlags[action.key] = val;
      return null;
    }

    /* ── 아이템 ── */
    if (act === 'add_item') {
      const count = action.count || 1;
      for (let i = 0; i < count; i++) {
        player.inventory.push(action.item);
      }
      if (action.message !== false) {
        UI.addSystemMsg(`  ▶ ${action.item} x${count}을(를) 획득했습니다!`);
      }
      return null;
    }

    if (act === 'remove_item') {
      const idx = player.inventory.indexOf(action.item);
      if (idx !== -1) player.inventory.splice(idx, 1);
      return null;
    }

    /* ── 골드 ── */
    if (act === 'add_gold') {
      const amt = action.amount;
      player.gold += amt;
      if (action.message !== false) {
        UI.addSystemMsg(`  ▶ ${amt}G 획득!`);
      }
      UI.updateHeader();
      return null;
    }

    if (act === 'sub_gold') {
      const amt = action.amount;
      player.gold = Math.max(0, player.gold - amt);
      if (action.message !== false) {
        UI.addSystemMsg(`  ▶ ${amt}G 소비!`);
      }
      UI.updateHeader();
      return null;
    }

    /* ── 스탯 ── */
    if (act === 'heal') {
      const amt = action.amount;
      if (amt === 'full' || amt === undefined || amt === null) {
        player.hp = player.maxHp;
      } else {
        player.heal(amt);
      }
      if (action.message !== false) {
        UI.addSystemMsg(`  ▶ HP가 회복되었습니다! (${player.hp}/${player.maxHp})`);
      }
      UI.updateHeader();
      return null;
    }

    if (act === 'damage') {
      const amt = action.amount;
      player.hp = Math.max(1, player.hp - amt);
      if (action.message !== false) {
        UI.addSystemMsg(`  ▶ ${amt}의 피해를 입었습니다! (HP: ${player.hp})`);
      }
      UI.updateHeader();
      return null;
    }

    if (act === 'add_stat') {
      for (const stat of ['attack', 'defense', 'max_hp']) {
        if (stat in action) {
          if (stat === 'max_hp') {
            player.maxHp += action[stat];
          } else {
            player[stat] += action[stat];
          }
          if (action.message !== false) {
            UI.addSystemMsg(`  ▶ ${stat} +${action[stat]}!`);
          }
        }
      }
      UI.updateHeader();
      return null;
    }

    if (act === 'set_stat') {
      for (const stat of ['attack', 'defense', 'max_hp', 'hp']) {
        if (stat in action) {
          if (stat === 'max_hp') {
            player.maxHp = action[stat];
          } else {
            player[stat] = action[stat];
          }
        }
      }
      UI.updateHeader();
      return null;
    }

    /* ── 어둠 점수 ── */
    if (act === 'add_dark') {
      const amt = action.amount || 1;
      player.darkPoints += amt;
      if (action.message !== false) {
        UI.addSystemMsg(`  ▶ 어둠 점수 +${amt}`);
      }
      return null;
    }

    if (act === 'sub_dark') {
      const amt = action.amount || 1;
      player.darkPoints = Math.max(0, player.darkPoints - amt);
      if (action.message !== false) {
        UI.addSystemMsg(`  ▶ 어둠 점수 -${amt}`);
      }
      return null;
    }

    if (act === 'set_dark') {
      player.darkPoints = action.amount;
      return null;
    }

    /* ── 전투 ── */
    if (act === 'battle') {
      const opts = {};
      if (action.name) opts.name = action.name;
      if (action.hp_multiply) opts.hp_multiply = action.hp_multiply;
      if (action.hp_add) opts.hp_add = action.hp_add;

      const result = await CombatSystem.startBattle(player, action.enemy, opts);

      if (result === 'lose') {
        const onLose = action.on_lose || [{ action: 'game_over' }];
        return await this.runActions(onLose, player, scenes);
      }
      if (result === 'fled') {
        return null;
      }
      const onWin = action.on_win || [];
      return await this.runActions(onWin, player, scenes);
    }

    /* ── 상점 ── */
    if (act === 'open_shop') {
      await UI.showShop(player, SHOP_STOCK, '상점');
      UI.showScreen('screen-game');
      return null;
    }

    if (act === 'open_merc_shop') {
      await UI.showShop(player, MERC_SHOP_STOCK, '용병 상점');
      UI.showScreen('screen-game');
      return null;
    }

    if (act === 'open_uw_shop') {
      await UI.showShop(player, UW_SHOP_STOCK, '암흑 시장');
      UI.showScreen('screen-game');
      return null;
    }

    if (act === 'open_cel_shop') {
      await UI.showShop(player, CEL_SHOP_STOCK, '천상 상점');
      UI.showScreen('screen-game');
      return null;
    }

    if (act === 'warp_map') {
      const targetMap = action.map;
      const targetZone = action.zone;
      const mapInfo = MAP_REGISTRY[targetMap];
      if (!mapInfo) return null;
      const locs = mapInfo.locations;
      const marker = Object.entries(locs).find(([, v]) => v.zone === targetZone);
      if (!marker) return null;
      // 마커 위치 찾기
      const raw = mapInfo.raw;
      let pos = null;
      for (let r = 0; r < raw.length && !pos; r++) {
        const idx = raw[r].indexOf(marker[0]);
        if (idx !== -1) pos = { row: r, col: idx };
      }
      if (!pos) return null;
      player.mapId = targetMap;
      player.mapRow = pos.row;
      player.mapCol = pos.col;
      player.currentLocation = targetZone;
      player.visitedLocations.add(targetZone);
      UI.updateHeader();
      UI.addSystemMsg(`  ★ ${mapInfo.name} - ${locs[marker[0]].name}(으)로 이동했습니다!`);
      return null;
    }

    /* ── 인벤토리/상태 ── */
    if (act === 'show_inventory') {
      await UI.showInventory(player);
      return null;
    }

    if (act === 'equip_item') {
      await UI.showEquip(player);
      return null;
    }

    if (act === 'show_status') {
      UI.showStatus(player);
      await UI.waitForTap();
      return null;
    }

    /* ── 위치 ── */
    if (act === 'set_location') {
      player.currentLocation = action.zone;
      player.visitedLocations.add(action.zone);
      return null;
    }

    /* ── 지도 ── */
    if (act === 'show_map') {
      UI.addLog('  [지도 표시 - 웹 버전에서는 위치 메뉴를 사용하세요]');
      return null;
    }

    /* ── 저장 ── */
    if (act === 'save') {
      UI.addDivider('게임 저장');
      const labels = [];
      for (let i = 0; i < 5; i++) labels.push(`슬롯 ${i + 1}`);
      labels.push('취소');
      const choice = await UI.showChoices(labels);
      if (choice < 5) {
        GameState.saveToLocal(choice);
        UI.addSystemMsg(`  슬롯 ${choice + 1}에 저장되었습니다!`);
        await UI.waitForTap();
      }
      return null;
    }

    /* ── 게임 오버 ── */
    if (act === 'game_over') {
      return 'gameover';
    }

    /* ── 반환 ── */
    if (act === 'return') {
      return action.value;
    }

    /* ── 선택지 메뉴 ── */
    if (act === 'menu') {
      return await this.runMenu(action, player, scenes);
    }

    /* ── 조건 분기 ── */
    if (act === 'if') {
      if (this.checkCondition(action.cond, player)) {
        return await this.runActions(action.then || [], player, scenes);
      } else {
        return await this.runActions(action.else || action['else'] || [], player, scenes);
      }
    }

    /* ── 랜덤 분기 ── */
    if (act === 'random') {
      return await this.runRandom(action, player, scenes);
    }

    /* ── 씬 이동 ── */
    if (act === 'goto') {
      const scene = scenes[action.scene];
      if (scene) {
        return await this.runScene(scene, player, scenes);
      }
      return null;
    }

    /* ── 반복 ── */
    if (act === 'loop') {
      while (true) {
        const result = await this.runActions(action.body || [], player, scenes);
        if (result !== null && result !== undefined) {
          if (result === '__break__') return null;
          return result;
        }
      }
    }

    if (act === 'break') {
      return '__break__';
    }

    return null;
  },

  /* ───── 메뉴 (선택지) ───── */

  async runMenu(action, player, scenes) {
    const optionsDef = action.options || [];

    const visible = optionsDef.filter(opt => {
      if (opt.when === undefined || opt.when === null) return true;
      return this.checkCondition(opt.when, player);
    });

    if (visible.length === 0) return null;

    const labels = visible.map(opt => this.formatText(opt.label, player));
    const choice = await UI.showChoices(labels);
    const chosen = visible[choice];

    if (chosen.goto) {
      const scene = scenes[chosen.goto];
      if (scene) return await this.runScene(scene, player, scenes);
      return null;
    }

    return await this.runActions(chosen.actions || [], player, scenes);
  },

  /* ───── 랜덤 분기 ───── */

  async runRandom(action, player, scenes) {
    const branches = action.branches || action.options || [];
    if (branches.length === 0) return null;

    const weights = branches.map(b => b.weight || 1.0);
    const total = weights.reduce((s, w) => s + w, 0);
    let r = Math.random() * total;
    let cumulative = 0;

    for (const branch of branches) {
      cumulative += branch.weight || 1.0;
      if (r < cumulative) {
        return await this.runActions(branch.actions || [], player, scenes);
      }
    }
    return await this.runActions(branches[branches.length - 1].actions || [], player, scenes);
  },

  /* ───── 씬 실행 ───── */

  async runScene(scene, player, scenes) {
    if (Array.isArray(scene)) {
      return await this.runActions(scene, player, scenes);
    }
    const actions = scene.actions || [];
    return await this.runActions(actions, player, scenes);
  },

  /* ───── YAML 파일 로드 + 실행 ───── */

  async runYamlEvent(player, zone) {
    const data = await GameState.loadYaml(zone);
    if (!data) return null;

    const event = data[zone] || data;
    const scenes = event.scenes || {};
    const mainActions = event.actions || [];

    return await this.runActions(mainActions, player, scenes);
  },

  /* ───── 지역 잠금 확인 ───── */

  isZoneLocked(player, zone) {
    const area = AREAS[zone];
    if (!area) return false;
    const cond = area.unlock_condition;
    if (!cond) return false;

    const flags = player.storyFlags;

    if (cond.flag) return !flags[cond.flag];
    if (cond.any_flag) return !cond.any_flag.some(f => !!flags[f]);

    return false;
  },

  getLockHint(zone) {
    const area = AREAS[zone];
    return area ? (area.lock_hint || '아직 이동할 수 없습니다.') : '알 수 없는 지역입니다.';
  },
};
